### DISCUSSION TITLE
Reload script file after editing?

### DISCUSSION THREAD

**Discussion by MickDuprez** [ClearFoundry/ClearScript#523] [Q&A] [answered]
Hi, search as I might I can't find any clear references on reloading a script in a running host application after making edits. Is there a method to load/reload a script after edits? 

I've tried 'Execute' on the same file but it seems it's not replacing it as expected/hoped.
I would like end users to be able to create there own scripts and load/reload on the fly while testing them.

When I call the sample code below everything works fine and modules are imported as expected, I make a change to one of the imported files, rerun this code but the changes are not being processed.

      engine.Execute(new DocumentInfo("Test.js") { Category = ModuleCategory.Standard }, @"
      var Editor = App.DocumentManager.MdiActiveDocument.Editor;
      Editor.WriteMessage('Javascript Engine is loaded!');
    
      import * as Test from 'test.js';
      Test.test();
      Test.test2();
    ");

Am I missing something simple? Thanks for any ideas, cheers.
Mick

**Comment by MickDuprez** [comment, marked as answer]
It's always the way, you find the solution after asking the question...

This seems to be working for now, cheers.

                DocumentLoader.Default.DiscardCachedDocuments();
                GC.Collect();
                GC.WaitForPendingFinalizers();

solution found here: https://github.com/microsoft/ClearScript/issues/187

Thanks

p.s.
There is also the DocumentLoader.CacheDocument Method that I need to look into for individual script file replacement, hth.

**Comment by MickDuprez** [comment]
hmm, CacheDocument is a no go at the moment, I'm getting an exception: 
  System.NullReferenceException: 'Object reference not set to an instance of an object.' 
from the CacheDocument call. Digging into the exception gives no further clues.

      public void JSReloadScript()
              {
                  var scriptsDir = Path.Combine(AppLoader.AppPath, "scripts");
                  var scriptPath = scriptsDir + "\\jstest.js";
                  string contents = File.ReadAllText(scriptPath);
                  var docInfo = new DocumentInfo("jsctest.js");
                  var strDoc = new StringDocument(docInfo, contents);
                  DocumentLoader.Default.CacheDocument(strDoc, true);
              }

Any suggestions on how to 'reload' a script file from the 'engine' or is there a way to do this in JavaScript code run by the engine (hot reload)?

The DiscardCachedDocuments works fine for the moment but I can see this being a problem once I have whole applications with lots of script files/modules running within the script engine.

**Reply by ClearScriptLib** [reply]
Hi @MickDuprez,

>I'm getting an exception: System.NullReferenceException: 'Object reference not set to an instance of an object.'

A cached document must have an absolute URI. If `scriptPath` is a full path, the following should work:

``` C#
var docInfo = new DocumentInfo(new Uri(scriptPath));
var strDoc = new StringDocument(docInfo, contents);
DocumentLoader.Default.CacheDocument(strDoc, true);
```

ClearScript is throwing the wrong exception – and we'll fix that in the next release – but the URI issue is most likely the problem.

Good luck!

**Comment by MickDuprez** [comment]
Thanks for the quick response!
Yes, that fixed the exception problem (and I remember reading somewhere about the absolute Uri, my bad :) ) but it doesn't seem to be the solution to reloading scripts. I'll keep digging, thanks again.
cheers
Mick
